logviewer/api/pod.go

```diff
@@ -0,0 +1,100 @@
+package api
+
+import (
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	namespace = "default"
+)
+
+type podLogsHandler struct {
+	clientset *kubernetes.Clientset
+}
+
+// ListLogsPodHandler is an HTTP handler to proxy /listlogs/:podid/... to the in-pod logviewer
+func (h *podLogsHandler) ListLogsPodHandler(w http.ResponseWriter, r *http.Request) {
+	podID, err := containerIDFromURL(r.URL.Path, listLogsExp)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+
+	h.proxyToPod(podID, w, r)
+}
+
+// LogPodHandler is an HTTP handler that proxies /logs/:podid/... to the in-pod log viewer
+func (h *podLogsHandler) LogPodHandler(w http.ResponseWriter, r *http.Request) {
+	podID, err := containerIDFromURL(r.URL.Path, logsExp)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+
+	h.proxyToPod(podID, w, r)
+}
+
+func (h *podLogsHandler) proxyToPod(podID string, w http.ResponseWriter, r *http.Request) {
+	pod, err := h.clientset.CoreV1().Pods(namespace).Get(r.Context(), podID, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	IP := pod.Status.PodIP
+	if IP == "" {
+		http.Error(w, "No IP found for the targetted pod", http.StatusNotFound)
+		return
+	}
+	url, err := url.Parse("http://" + IP + ":8004")
```

Comment:
How will you make sure you have connectivity to the pod IP, and the titus agent isn't filtered from reaching it? For example, for the studio accounts, it is this way. @rgulewich wrote a reverse proxy to solve this issue that goes across namespaces.

Reply:
Thanks for the heads up, I will reach out to @rgulewich about that use case.  I ran this past @alekar but am open to your thoughts on how we should route internally.  An alternatives could be a host accessible mount with named pipe created by the log-viewer at startup.  I think this would work, but would have to try before I was confident but maybe worth investigation.

Reply:
I would prefer unix socket over pipe. But yeah, either approach works.

Reply:
The socket approach is a good idea.  What do you think about preserving the current way of proxying, and writing an OCI hook to do the bind mounting of the container's pid 1, like we do here - https://github.com/Netflix/titus-executor/blob/master/executor/runtime/docker/docker_linux.go#L146